ostree/ostree_dest.go

```diff
@@ -32,7 +33,7 @@ import (
 	"github.com/vbatts/tar-split/tar/storage"
 )
 
-// #cgo pkg-config: glib-2.0 gobject-2.0 ostree-1 libselinux
```

Comment:
The inclusion of the headers enables loading `libselinux`, it only makes the constants and functions known. In fact **not including the header would lead to build errors**.

The difference between using `pkg-config` and `dlopen` to load `libselinux` is when you are getting a **runtime error**.

## Runtime behavior

### Current code
- If building without the flag `selinux`, github.com/opencontainers/selinux gets built to [return `false` to the selinux check](https://github.com/opencontainers/selinux/blob/master/go-selinux/selinux_stub.go#L34).
- If building with the flag `selinux`, github.com/opencontainers/selinux gets built so that [the function `GetEnabled` returns `true` only if SELinux is effectively enabled](https://github.com/opencontainers/selinux/blob/51c6c0a5dbc675792e953298cb9871819d6f9bb8/go-selinux/selinux_linux.go#L204).

In any case since `pkg-config` automatically tries to load `libselinux`, **you get an error if it's missing**. Therefore with the current code **`libselinux` has to be present if building with OStree support**, no matter the status of the buildflag `selinux`.

- If building without the flag `selinux`, `libselinux` has to be present, but won't be used!
- If building with the flag `selinux`, `libselinux` has to be present, and will be effectibly used.

The value of `selinuxHnd` is guarding any code involving SELinux from getting executed, if it's not enabled. And it happens independently from the `selinux` build flag.

### Patch 

The usage of [`dlopen`](https://github.com/containers/image/pull/546/files#diff-952ad9dc01673ce7143c20dd5808adf7R525) tries to load `libselinux`, but **doesn't return error if `libselinux` is missing**.

It does return **error if `libselinux` is expected to get loaded and cannot be loaded**. But what does it "expected to get loaded" means?

If it was built with the flag `selinux` and SELinux is enabled ([GetEnabled](https://github.com/opencontainers/selinux/blob/51c6c0a5dbc675792e953298cb9871819d6f9bb8/go-selinux/selinux_linux.go#L204) returns `true`).

In any case the patch doesn't change the guarding behavior of the value of `selinuxHnd` to avoid executing SELinux code, if not needed.

*Sorry for the long explanation, but I think that it will be helpful to understand the expected behavior*.

Reply:
This is another reason why actually using the `selinux` build tag instead of run-time `dlopen` would be better:

If the SELinux consumers were gated behind a build tag, a non-`selinux` build could succeed without `libselinux` headers and development libraries being present on the build system at all.  Considering how many C projects use something like autoconf and change their build options depending on whether the development files for `libselinux` are present, not including them in the build environment seem quite desirable.

And if you are considering a special-case embedded system, it would certainly be nice not to have to build `libselinux` for the target platform at all, if you don’t plan to use it!